prov/gni/src/gnix_cm.c

```diff
@@ -53,122 +111,787 @@
+			return ret;
 		}
 
 		break;
-	case FI_CLASS_SEP:
-		sep_priv = container_of(fid, struct gnix_fid_sep, ep_fid);
-		name = sep_priv->my_name;
-		break;
+	case GNIX_PEP_SOCK_RESP_REJECT:
+		GNIX_INFO(FI_LOG_EP_CTRL, "reject not implemented\n");
+		return -FI_EINVAL;
 	default:
+		GNIX_INFO(FI_LOG_EP_CTRL, "Invalid response command: %d\n",
+			  resp->cmd);
+		return -FI_EINVAL;
+	}
+
+	return FI_SUCCESS;
+}
+
+/* Check for a connection response on an FI_EP_MSG. */
+int _gnix_ep_progress(struct gnix_fid_ep *ep)
+{
+	int ret, bytes_read;
+	struct gnix_pep_sock_connresp resp;
+
+	COND_ACQUIRE(ep->requires_lock, &ep->vc_lock);
+
+	/* Check for a connection response. */
+	bytes_read = read(ep->conn_fd, &resp, sizeof(resp));
+	if (bytes_read == sizeof(resp)) {
+		/* Received response. */
+		ret = __gnix_ep_connresp(ep, &resp);
+		if (ret != FI_SUCCESS) {
+			GNIX_WARN(FI_LOG_EP_CTRL,
+				  "__gnix_pep_connreq failed, err: %d\n",
+				  ret);
+		}
+	} else if (errno != EAGAIN) {
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			  "Unexpected errno on listen socket: %d\n",
+			  errno);
+	}
+
+	COND_RELEASE(ep->requires_lock, &ep->vc_lock);
+
+	return FI_SUCCESS;
+}
+
+DIRECT_FN STATIC int gnix_connect(struct fid_ep *ep, const void *addr,
+				  const void *param, size_t paramlen)
+{
+	int ret;
+	struct gnix_fid_ep *ep_priv;
+	struct sockaddr_in saddr;
+	struct gnix_pep_sock_connreq req;
+	struct gnix_vc *vc;
+	struct gnix_mbox *mbox = NULL;
+	struct gnix_av_addr_entry av_entry;
+
+	if (!ep || !addr)
+		return -FI_EINVAL;
+
+	ep_priv = container_of(ep, struct gnix_fid_ep, ep_fid.fid);
+
+	COND_ACQUIRE(ep_priv->requires_lock, &ep_priv->vc_lock);
+
+	if (ep_priv->conn_state != GNIX_EP_UNCONNECTED) {
 		ret = -FI_EINVAL;
-		goto err;
+		goto err_unlock;
+	}
+
+	ret = gnixu_pe_to_ip(addr, &saddr);
+	if (ret != FI_SUCCESS) {
+		GNIX_INFO(FI_LOG_EP_CTRL,
+			  "Failed to translate gnix_ep_name to IP\n");
+		goto err_unlock;
+	}
+
+	/* Create new VC without CM data. */
+	av_entry.gnix_addr = ep_priv->dest_addr.gnix_addr;
+	av_entry.cm_nic_cdm_id = ep_priv->dest_addr.cm_nic_cdm_id;
+	ret = _gnix_vc_alloc(ep_priv, &av_entry, &vc);
+	if (ret != FI_SUCCESS) {
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			  "Failed to create VC:: %d\n",
+			  ret);
+		goto err_unlock;
 	}
+	ep_priv->vc = vc;
 
-	/*
-	 * Retrieve the cdm_id & device_addr from the gnix_cm_nic structure.
-	 */
+	ret = _gnix_mbox_alloc(vc->ep->nic->mbox_hndl, &mbox);
+	if (ret != FI_SUCCESS) {
+		GNIX_WARN(FI_LOG_EP_DATA,
+			  "_gnix_mbox_alloc returned %s\n",
+			  fi_strerror(-ret));
+		goto err_mbox_alloc;
+	}
+	vc->smsg_mbox = mbox;
 
-	memcpy(addr, &name, copy_size);
+	ep_priv->conn_fd = socket(AF_INET, SOCK_STREAM, 0);
+	if (ep_priv->conn_fd < 0) {
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			  "Failed to create connect socket, errno: %d\n",
+			  errno);
+		ret = -FI_ENOSPC;
+		goto err_socket;
+	}
+
+	/* Currently blocks until connected. */
+	ret = connect(ep_priv->conn_fd, (struct sockaddr *)&saddr,
```

Comment:
Yes that can be improved by making the socket non-blocking and completing the gnix_connect() work during EQ progress.